007-design-docs.md

```diff
@@ -0,0 +1,124 @@
+- If just a couple of named contributors moved on to other things, the velocity of the project would drop considerably. Design doc would help in this case too, capturing conventions and behaviour enforcements that normally only get picked up and tweaked at review time.
+
+
+
+## Where is Strimzi today
+
+Currently there are no design documents for Strimzi, the closest thing Strimzi has would probably be the generated docs for classes - these are more user oriented than design oriented. Some classes also have extensive documentation comments found at the top of their files explaining their usage and design.
+
+
+## Proposal
+
+### When to write a design document
+
+It would be appropriate to add a design document when anything which touches a 'Public API'.
+Any of the following should be considered a "public API":
+- Additional Custom Resource Definitions (CRD) or structure changes of any of the project's CRDs (excluding documentation of field)
+- Changes in how CRDs are used or function (e.g. annotation changes)
+- Anything which would add, remove or change any other public Strimzi APIs (e.g. New endpoints in the bridge, or new REST APIs).
+- Anything which changes the accessibility of any of Kafka's public APIs (e.g. changes to how metrics get exposed, changing which broker configs are reserved to the operators, or locking/exposing functionality either via the network, or Kafka protocol, such as authorization)
+
+
+### Where to write a design document
+
+A design document should be a markdown file located in the `design` directory of the repository they are located in, for instance all design documents relating to the strimzi operator(s) should be located in the `strimzi-kafka-operator` repository under the `design` folder.
+
+#### Strimzi Design Document Organization
+Design documents should be splt into logical areas, for repositories that aren't `strimiz-kafka-operator` I suspect this would start as a single or just a few succinct documents.
+For `strimzi-kafka-operator` I would suggest splitting design documentation from 'public API' perspective, that is to say all documents relating to the `Kafka` custom resource would be separated into its own folder called `Kafka`, similarly for `KafkaConnect` etc.
+For design documentation that would cover the whole operator and not just any single custom resource I would suggest a folder named `operator` this would describe behaviours not unique to any one specific reconcile loop or verticle.
+
+
+### When is a design document complete
+
+A design document is considered complete when the reviewers and maintainers believe the document to accurately reflect the current implementation of the components behaviour. Once approved the document is merged and is understood to be correct, any mistakes or changes to behaviour should be corrected in the form of a pull request.
```

Comment:
I do not think this really addressed my previous comment. I think this has again very confusing boundaries between proposals and your design docs. As I read it, I would need to:
1) Write a proposal and get it approved
2) Write a design docs and get it approved
3) Do the actual code (this is where you figure out the proposal and design docs do not work :-o)

This would totally stop any progress. As I suggested before, I think the way this can be done is that the design docs are technical design documentation written as part of the code PR. But I guess there might be other options. But I do not think it is possible to build another approval barrier.

As an example, we need to for example learn from @ppatierno's proposal around bridge metrics. Which shows that we maybe didn't yet learn properly how to handle the proposals before introducing some new hurdle.